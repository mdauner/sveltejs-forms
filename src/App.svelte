<script>
  import {
    Form,
    Input,
    Select,
    Choice,
  } from './components/components.module.js';
  import * as yup from 'yup';

  function handleReset() {
    console.log('form has been reset');
  }

  function handleSubmit({ detail: { values, setSubmitting, resetForm } }) {
    setTimeout(() => {
      console.log(values);
      setSubmitting(false);
      resetForm({
        user: { email: 'test@user.com' },
      });
    }, 2000);
  }

  const schema = yup.object().shape({
    user: yup.object().shape({
      email: yup
        .string()
        .required()
        .email(),
      password: yup.string().min(4),
      language: yup.string().required(),
      os: yup
        .array()
        .of(yup.string().required())
        .min(2),
    }),
  });

  const langOptions = [
    { id: 'svelte', title: 'Svelte' },
    { id: 'react', title: 'React' },
    { id: 'angular', title: 'Angular' },
  ];

  const osOptions = [
    { id: 'macos', title: 'macOS' },
    { id: 'linux', title: 'Linux üêß' },
    { id: 'windows', title: 'Windows' },
  ];
</script>

<style lang="scss">
  :global(.sveltejs-forms) {
    background-color: lightgray;
    display: inline-block;
    padding: 1rem;
    .field {
      margin-bottom: 1rem;
      &.error {
        input,
        select {
          border: 1px solid red;
        }
        .message {
          margin-top: 0.2rem;
          color: red;
          font-size: 0.8rem;
        }
      }
    }
  }
</style>

<Form
  {schema}
  validateOnChange={true}
  validateOnBlur={true}
  on:submit={handleSubmit}
  on:reset={handleReset}
  let:isSubmitting
  let:isValid>
  <Input name="user.email" placeholder="Email" />
  <Input
    name="user.password"
    type="password"
    placeholder="Password"
    multiline />
  <Select name="user.language" options={langOptions} />
  <Choice name="user.os" options={osOptions} multiple />

  <button type="reset">Reset</button>
  <button type="submit" disabled={isSubmitting}>Sign in</button>
  The form is valid: {isValid}
</Form>
