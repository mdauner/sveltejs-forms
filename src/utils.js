export function createObjectWithDefaultValue(defaultValue = '') {
  return new Proxy(
    {},
    {
      get: function(object, property) {
        return Object.prototype.hasOwnProperty.call(object, property)
          ? object[property]
          : defaultValue;
      },
    }
  );
}
