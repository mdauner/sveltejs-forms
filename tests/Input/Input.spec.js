import App from './TestApp.svelte';
import { fireEvent, wait } from '@testing-library/svelte';
import * as yup from 'yup';
import { render } from '../utils';

describe('Input', () => {
  it('renders textarea when multiline parameter is set', async () => {
    const { container } = await render(App, { props: { multiline: true } });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('updates form value on change', async () => {
    const { component, getByPlaceholderText } = await render(App);

    const emailInput = getByPlaceholderText('Email');

    await fireEvent.change(emailInput, {
      target: { value: 'test@user.com' },
    });

    expect(component.form.$$.ctx.$values).toMatchObject({
      email: 'test@user.com',
    });
  });

  it('validates on change if validateOnChange is true', async () => {
    const schema = yup.object().shape({
      email: yup.string().email(),
    });
    const { component, getByPlaceholderText } = await render(App, {
      props: { schema, validateOnChange: true },
    });

    const emailInput = getByPlaceholderText('Email');

    await fireEvent.change(emailInput, {
      target: { value: 'invalid value' },
    });
    await wait();

    expect(component.form.$$.ctx.$errors).toEqual({
      email: 'email must be a valid email',
    });
  });

  it('does not validate on change if validateOnChange is false', async () => {
    const schema = yup.object().shape({
      email: yup.string().email(),
    });
    const { component, getByPlaceholderText } = await render(App, {
      props: { schema, validateOnChange: false },
    });

    const emailInput = getByPlaceholderText('Email');

    await fireEvent.change(emailInput, {
      target: { value: 'invalid value' },
    });
    await wait();

    expect(component.form.$$.ctx.$errors).toEqual({});
  });

  it('validates on blur if validateOnBlur is true', async () => {
    const schema = yup.object().shape({
      email: yup.string().email(),
    });
    const { component, getByPlaceholderText } = await render(App, {
      props: { schema, validateOnChange: false, validateOnBlur: true },
    });

    const emailInput = getByPlaceholderText('Email');

    await fireEvent.change(emailInput, {
      target: { value: 'invalid value' },
    });
    await fireEvent.blur(emailInput);
    await wait();

    expect(component.form.$$.ctx.$errors).toEqual({
      email: 'email must be a valid email',
    });
  });

  it('does not validate on blur if validateOnBlur is false', async () => {
    const schema = yup.object().shape({
      email: yup.string().email(),
    });
    const { component, getByPlaceholderText } = await render(App, {
      props: { schema, validateOnChange: false, validateOnBlur: false },
    });

    const emailInput = getByPlaceholderText('Email');

    await fireEvent.change(emailInput, {
      target: { value: 'invalid value' },
    });
    await fireEvent.blur(emailInput);
    await wait();
    expect(component.form.$$.ctx.$errors).toEqual({});
  });

  it('matches snapshot', async () => {
    const { container } = await render(App);
    expect(container.firstChild).toMatchSnapshot();
  });
});
