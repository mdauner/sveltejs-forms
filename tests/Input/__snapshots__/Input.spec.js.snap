// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches snapshot 1`] = `
<div>
  <form
    class="sveltejs-forms"
  >
    <div
      class="field"
    >
      <input
        name="email"
        placeholder="Email"
        type="text"
      />
       
    </div>
     
    <button
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;

exports[`Input renders textarea when multiline parameter is set 1`] = `
<div>
  <form
    class="sveltejs-forms"
  >
    <div
      class="field"
    >
      <textarea
        name="email"
        placeholder="Email"
      />
       
    </div>
     
    <button
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;
